## WF2 Module 2

We want to find molecular phenotypes that are integral to a disease, and then chemicals that activate those phenotypes, and then genes that clean up those bad chemicals.   The tricky part with current data is that it's difficult to make the jump to chemicals.

For instance, if we say that molecular phenotype is a GO process, we might find that e.g. various DNA repair processes are involved with Fanconi Anemia.  But the only chemicals that are linked to these processes are participants in the process, like DNA.

From the chemical side, there are properties on chemicals in CHEBI, such as being genotoxic.  However, there's not any semantic value to these properties because they are just strings with no relation to e.g. GO terms.  This gap is being filled by CHIRO, but that is still ongoing.

To leapfrog this, we'll try (1) grabbing chemicals that look bad for the disease, (2) try to do some enrichments on their properties to get other similar chemicals, and (3) look for genes that degrade those things.

In [1]:
#To make nicer looking outputs
from IPython.core.display import display, HTML
import requests
import pandas as pd
import os
import sys

#Load some functions for parsing quick output
module_path = os.path.abspath(os.path.join('../..'))
if module_path not in sys.path:
        sys.path.append(module_path)
from gg_functions import parse_answer, get_view_url, expand, quick

robokop='robokop.renci.org' 

In [2]:
disease = "MONDO:0019391"  #Fanconi Anemia

In [3]:
bad_chemicals = expand('disease',disease,'chemical_substance',predicate='contributes_to',direction='in')

In [5]:
df = parse_answer(bad_chemicals,node_list=['n1'])

In [6]:
df

n1 - id                n1 - name     score
0      CHEBI:9288             streptozocin  1.390503
1     CHEBI:76451                  alloxan  1.092735
2     CHEBI:22977             cadmium atom  0.963028
3      CHEBI:8382               prednisone  0.919533
4      CHEBI:7735               olanzapine  0.852368
5     CHEBI:18248                iron atom  0.840737
6     CHEBI:17158            methylglyoxal  0.780709
7      CHEBI:3766                clozapine  0.756415
8     CHEBI:28694              copper atom  0.751212
9     CHEBI:16598                      DDE  0.751212
10    CHEBI:82160                   Fenson  0.740737
11    CHEBI:47217                   arsane  0.740737
12    CHEBI:61057       tacrolimus hydrate  0.740737
13   MESH:D052638       Particulate Matter  0.740737
14    CHEBI:16503                   selane  0.651212
15     CHEBI:5391                 glucagon  0.644814
16     CHEBI:8378             prednisolone  0.642490
17    CHEBI:33216              bisphenol A  0.640111
18    CHEBI:23808    dimethylarsinous acid  0.640111
19    CHEBI:25812                    ozone  0.640111
20    CHEBI:28685          molybdenum atom  0.640111
21    CHEBI:38764                  phorate  0.640111
22    CHEBI:27998                 tungsten  0.640111
23    CHEBI:26523  reactive oxygen species  0.628686
24    CHEBI:24261           glucocorticoid  0.547317
25     CHEBI:6397                     Lead  0.540111
26   MESH:D001152               Arsenicals  0.540111
27   MESH:D000305  Adrenal Cortex Hormones  0.540111
28     CHEBI:4031            cyclosporin A  0.519937
29     CHEBI:4495                diazoxide  0.518851
..            ...                      ...       ...
220   CHEBI:10650              sumatriptan  0.359326
221    CHEBI:3897        Cortisone acetate  0.359326
222  CHEBI:135361          methscopolamine  0.359326
223    CHEBI:3437                carteolol  0.359326
224   CHEBI:31225               antipyrine  0.359326
225    CHEBI:6909            metronidazole  0.359326
226   CHEBI:50275             esomeprazole  0.359326
227    CHEBI:2639                amiloride  0.359326
228   CHEBI:17115                 L-serine  0.359326
229    CHEBI:7822                oxaprozin  0.359326
230    CHEBI:4798               entacapone  0.359326
231    CHEBI:3077            betamethasone  0.359326
232    CHEBI:3561               cetirizine  0.359326
233    CHEBI:6504        lithium carbonate  0.359326
234    CHEBI:6824   hexamethylenetetramine  0.359326
235   CHEBI:63633              triptorelin  0.359326
236   CHEBI:49575                 diazepam  0.359326
237   CHEBI:36773                  camphor  0.359326
238   CHEBI:71416          phenazopyridine  0.359326
239   CHEBI:31521            doxifluridine  0.359326
240    CHEBI:7731                ofloxacin  0.359326
241   CHEBI:84115               antazoline  0.359326
242    CHEBI:7601               nizatidine  0.359326
243    CHEBI:8887                rofecoxib  0.359326
244   CHEBI:28487                reserpine  0.359326
245   CHEBI:52717               bortezomib  0.359326
246    CHEBI:5739                histrelin  0.359326
247   CHEBI:16134                  ammonia  0.359326
248   CHEBI:16643             L-methionine  0.359326
249    CHEBI:4702              dorzolamide  0.359326

[250 rows x 3 columns]

What properties do these things have in common?

In [7]:
chems = [x for x in bad_chemicals['knowledge_graph']['nodes'] if 'chemical_substance' in x['type']]
from collections import defaultdict
prop_counts = defaultdict(int)
for n in chems:
    for p in n:
        prop_counts[p] += 1
items = list(prop_counts.items())
count_frame = pd.DataFrame.from_records(items,columns=['property','count_in_subset'])
count_frame.head()

property  count_in_subset
0                      id              250
1                    name              250
2  equivalent_identifiers              250
3                    type              250
4  omnicorp_article_count              250

Many of these propertries are assigned to many chemicals (like SMILES or name).  We want to find the ones that are overrepresented.  To do that we need to know the details of how often each property is used in the database.  At the moment, ROBOKOP does not expose functions to allow this, but we have pre-calculated such a list:

In [9]:
property_counts = pd.read_csv('../../examples/chemprops.txt','\t')
property_counts.head()

property  count
0                                  appetite_enhancer      2
1         Wiskott_Aldrich_syndrome_protein_inhibitor      1
2                           peptide_coupling_reagent      2
3                 EC_3.4.22.52_(calpain_1)_inhibitor      3
4  EC_2.3.1.5_(arylamine_N_acetyltransferase)_inh...      1

In [11]:
df = pd.merge(count_frame, property_counts,on='property',how='inner')

df.head()

property  count_in_subset   count
0                      id              250  355412
1                    name              250  311887
2  equivalent_identifiers              250  355412
3        therapeutic_flag              208  244803
4                 topical              208  244803

In [13]:
from scipy.stats import hypergeom
#in scipy:
#The hypergeometric distribution models drawing objects from a bin. 
#M is the total number of objects, 
#n is total number of Type I objects. 
#The random variate represents the number of Type I objects in N drawn without replacement from the total population.

total_chemical_count=355413 # M above
subset_count = len(chems) # N above

def calc_enrich_p(x,n):
    return hypergeom.sf(x-1, total_chemical_count, n, subset_count)

In [14]:
df['enrichment_p'] = df.apply(lambda x: calc_enrich_p(x['count_in_subset'], x['count']), axis=1)
df.sort_values(by='enrichment_p')

property  count_in_subset   count   enrichment_p
22                      application              162    8162  1.247381e-198
52                             drug              141    5527  4.179293e-184
45                   pharmaceutical              141    5605  3.018204e-183
57                  biological_role              184   27273  5.710007e-147
33             pharmacological_role               90    2313  4.849076e-129
5                              mass              219   97320   1.072944e-88
17                           charge              219   98675   1.891211e-87
6                 monoisotopic_mass              218   98074   9.602293e-87
71                    chemical_role              106   16524   2.391485e-72
42           neurotransmitter_agent               46     978   1.495992e-68
49                 enzyme_inhibitor               55    2224   9.875413e-67
46                        inhibitor               55    2445   1.706314e-64
144             cardiovascular_drug               36     416   1.962314e-63
95        environmental_contaminant               35     451   5.697158e-60
143          antihypertensive_agent               28     196   6.536746e-56
97                       xenobiotic               35     613   3.448501e-55
26     EC_3.*_(hydrolase)_inhibitor               37     814   1.336967e-54
44                           entity               78   13726   8.300056e-48
56                       continuant               78   13726   8.300056e-48
28                             role               78   13726   8.300056e-48
25                 biochemical_role               92   22692   4.867505e-45
41                 adrenergic_agent               24     251   1.213726e-43
102                         agonist               32     915   1.229307e-43
31                       antagonist               27     442   1.481822e-43
51               tranquilizing_drug               21     135   5.034633e-43
23                psychotropic_drug               23     257   3.514687e-41
70                         acceptor               57    7542   7.049886e-41
58      central_nervous_system_drug               26     478   1.209329e-40
68                             base               56    7531   8.931719e-40
69                    Bronsted_base               55    7524   1.139676e-38
..                              ...              ...     ...            ...
318                       mycotoxin                1      86   5.872650e-02
337                  anticonvulsant                1      90   6.137275e-02
338                     anaesthetic                1      98   6.664304e-02
381  platelet_aggregation_inhibitor                1     105   7.123036e-02
358    marine_xenobiotic_metabolite                1     107   7.253690e-02
317          Penicillium_metabolite                1     108   7.318948e-02
207                 antifungal_drug                1     109   7.384160e-02
247    human_blood_serum_metabolite                1     109   7.384160e-02
371                         solvent                1     109   7.384160e-02
256              hypoglycemic_agent                1     112   7.579523e-02
416                   plant_hormone                1     153   1.020871e-01
383               hematologic_agent                1     159   1.058717e-01
335                       indicator                1     190   1.251737e-01
412          plant_growth_regulator                1     194   1.276339e-01
404                         mutagen                1     195   1.282478e-01
283          Aspergillus_metabolite                1     202   1.325335e-01
406         antifungal_agrochemical                1     231   1.500662e-01
401                       genotoxin                1     237   1.536494e-01
413                growth_regulator                1     249   1.607706e-01
19                         inchikey              227  315397   1.768640e-01
11                            inchi              227  315397   1.768640e-01

Now we can see that the most enriched property is "mutagen", which occurs in 2 of the 3 chemicals that are contributors to FA, but in only 195 out of the 355k chemicals.  Let's look for genes that degrade mutagens.  We're going to lose the connection to FA, so that won't be included in our score, which is unfortunate.

In [15]:
def create_question(disease_id,prop):
    return {
    "machine_question": {
        "nodes": [
            {
                "id": "n0",
                "type": "chemical_substance",
                prop: True
            },
            {
                "id": "n1",
                "type": "gene"
            },
            {
                "id": "n2",
                "type": "disease",
                "curie": disease_id
            }
        ],
        "edges": [
            {
                "id": "e0",
                "source_id": "n1",
                "target_id": "n0",
                "type": "increases_degradation_of"
            }
        ]
    }
}

In [16]:
q = create_question('MONDO:0019391','mutagen')
a = quick(q)

Return Status: 200


In [17]:
aframe = parse_answer(a,node_list=['n0','n1'])

In [18]:
aframe

n0 - id               n0 - name     n1 - id n1 - name     score
0   CHEBI:17754                glycerol   HGNC:4289        GK  0.283601
1   CHEBI:28262      dimethyl sulfoxide   HGNC:4289        GK  0.274309
2   CHEBI:27732                caffeine   HGNC:2596    CYP1A2  0.266359
3   CHEBI:17754                glycerol    HGNC:381    AKR1B1  0.265992
4   CHEBI:17754                glycerol    HGNC:382   AKR1B10  0.264636
5   CHEBI:17754                glycerol  HGNC:28635       GK5  0.264351
6   CHEBI:15343            acetaldehyde    HGNC:404     ALDH2  0.263787
7    CHEBI:4027        cyclophosphamide   HGNC:2637    CYP3A4  0.257238
8    CHEBI:5864              ifosfamide   HGNC:2637    CYP3A4  0.257073
9   CHEBI:15343            acetaldehyde    HGNC:407   ALDH1B1  0.256739
10  CHEBI:15343            acetaldehyde    HGNC:877   ALDH7A1  0.256574
11  CHEBI:28262      dimethyl sulfoxide    HGNC:380    AKR1A1  0.256561
12  CHEBI:15343            acetaldehyde    HGNC:409   ALDH1A3  0.256559
13  CHEBI:15343            acetaldehyde    HGNC:412   ALDH9A1  0.256547
14  CHEBI:15343            acetaldehyde    HGNC:410   ALDH3B1  0.256533
15  CHEBI:17754                glycerol    HGNC:380    AKR1A1  0.226516
16  CHEBI:28262      dimethyl sulfoxide    HGNC:381    AKR1B1  0.220641
17  CHEBI:27561                 oxirane   HGNC:3402     EPHX2  0.219411
18  CHEBI:28262      dimethyl sulfoxide    HGNC:382   AKR1B10  0.219373
19  CHEBI:27561                 oxirane  HGNC:23760     EPHX3  0.219090
20  CHEBI:28262      dimethyl sulfoxide  HGNC:28635       GK5  0.219088
21  CHEBI:15343            acetaldehyde    HGNC:403   ALDH3A2  0.219069
22  CHEBI:17754                glycerol   HGNC:4291       GK2  0.187528
23  CHEBI:15343            acetaldehyde    HGNC:405   ALDH3A1  0.180899
24  CHEBI:28534       1,2-dibromoethane   HGNC:4641     GSTT1  0.180865
25   CHEBI:2560         Aldophosphamide    HGNC:409   ALDH1A3  0.180746
26  CHEBI:84254   N,N-dimethylacetamide   HGNC:4289        GK  0.160021
27  CHEBI:84254   N,N-dimethylacetamide    HGNC:381    AKR1B1  0.146686
28  CHEBI:84254   N,N-dimethylacetamide    HGNC:382   AKR1B10  0.145596
29  CHEBI:80559         Alcophosphamide    HGNC:250     ADH1B  0.145492
..          ...                     ...         ...       ...       ...
64   CHEBI:2560         Aldophosphamide   HGNC:4638     GSTP1  0.145334
65   CHEBI:2560         Aldophosphamide   HGNC:4632     GSTM1  0.145334
66  CHEBI:28534       1,2-dibromoethane   HGNC:4626     GSTA1  0.145334
67  CHEBI:15343            acetaldehyde    HGNC:411   ALDH3B2  0.145334
68  CHEBI:28262      dimethyl sulfoxide   HGNC:4291       GK2  0.145334
69  CHEBI:40154         5-fluorouridine  HGNC:12562      UCK2  0.121300
70   CHEBI:2560         Aldophosphamide    HGNC:405   ALDH3A1  0.121300
71  CHEBI:40154         5-fluorouridine  HGNC:14859      UCK1  0.121287
72  CHEBI:28534       1,2-dibromoethane  HGNC:23064     GSTO2  0.087940
73   CHEBI:2560         Aldophosphamide   HGNC:7063     MGST2  0.087940
74  CHEBI:80559         Alcophosphamide    HGNC:255      ADH6  0.087940
75  CHEBI:28534       1,2-dibromoethane   HGNC:4637     GSTM5  0.087940
76  CHEBI:80559         Alcophosphamide    HGNC:256      ADH7  0.087940
77  CHEBI:28534       1,2-dibromoethane  HGNC:19662     GSTA5  0.087940
78  CHEBI:84254   N,N-dimethylacetamide   HGNC:4291       GK2  0.087940
79  CHEBI:28534       1,2-dibromoethane  HGNC:33437    GSTT2B  0.087940
80  CHEBI:80559         Alcophosphamide    HGNC:249     ADH1A  0.087940
81  CHEBI:40154         5-fluorouridine  HGNC:23061      UPP2  0.087940
82   CHEBI:2560         Aldophosphamide  HGNC:17890     HPGDS  0.087940
83   CHEBI:2560         Aldophosphamide  HGNC:19662     GSTA5  0.087940
84  CHEBI:40154         5-fluorouridine  HGNC:12576      UPP1  0.087940
85   CHEBI:2560         Aldophosphamide  HGNC:23064     GSTO2  0.087940
86  CHEBI:76565  2-ammonioprop-2-enoate  HGNC:30404      SDSL  0.087940
87   CHEBI:2560         Aldophosphamide   HGNC:4

This finds the expected result that ALDH2 is a potential modifier for FA.   Potentially, but looking for things that degrade into acetaldehyde or other chemicals in n0 we could find a list of environemntal exposures to avoid.  Note that, as ever, we are at the mercy of the data - caffeine is mutagenic at high dose but mostly in non-mammalian cells.